### PR TITLE
Feat: add support for ALTER COLUMN DROP NOT NULL

### PR DESCRIPTION
**Pull request by noklam**
fix: #3534 
Context:
```python
class AlterColumn(Expression):
    arg_types = {
        "this": True,
        "dtype": False,
        "collate": False,
        "using": False,
        "default": False,
        "drop": False,
        "comment": False,

    }
```

The existing `AlterColumn` expression seems to assume only `SET DEFAULT ` and `NOT DEFAULT` is possible. Therefore is `drop` is True is sufficient to represent `DROP DEFAULT` and when `default` is not null it means set default. This seems to be insufficient to handle statement like `DROP NOT NULL` and `SET NOT NULL`

Few questions:
-  there are many `_match_xxx` functions, is `_match_text_seq` the most suitable one?
- I don't know if I should use things like `_parse_not_constraints` here.
- I figured I should edit the `AlterColumn` expression, but I am not sure how to add a new type for `SET NOT NULL` and `DROP NOT NULL`. I found there are `NotNullColumnConstrains`. I take a look at other expression and I thought adding a new optional arg_type `expression` seems to best fit.

# Test
Reuse the original issue example.

This is expected to work
```python
ast = parse_one(sql, read="spark")
ast.sql(write="duckdb")
```

Duckdb Syntax: https://duckdb.org/docs/sql/statements/alter_table.html
Snowflake Syntax: https://docs.snowflake.com/en/sql-reference/sql/alter-table-column

### CODE REVIEW

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -1632,6 +1632,7 @@ class AlterColumn(Expression):
         "default": False,
         "drop": False,
         "comment": False,
+        "expression": False
```
Comment (georgesittas) [inline sqlglot/expressions.py:1635]: Let's rename this to `not_null` so it's a bit clearer what it stands for.

File: sqlglot/parser.py [hunk 2]
```diff
@@ -6055,7 +6055,15 @@ def _parse_alter_table_alter(self) -> t.Optional[exp.Expression]:
             return self.expression(exp.AlterColumn, this=column, default=self._parse_conjunction())
         if self._match(TokenType.COMMENT):
             return self.expression(exp.AlterColumn, this=column, comment=self._parse_string())
-
+        if self._match_text_seq("DROP", "NOT" ,"NULL"):
+            return self.expression(exp.AlterColumn,
+                                   this=column,
+                                drop=True,
+                                   expression=exp.NotNullColumnConstraint)
```
Comment (georgesittas) [inline sqlglot/parser.py:6062]: This needs to be `expression=exp.NotNullColumnConstraint()`, because it currently assigns a class to the arg.

**Review by georgesittas — COMMENTED** [review]
Thanks for the PR @noklam, this is in the right direction - left a couple of comments. Make sure to also run `make style` and add tests in `test_snowflake.py`.

File: sqlglot/expressions.py (hunk 1 above)
Comment (noklam) [reply sqlglot/expressions.py:1635]: I will use `allow_null` instead as it seems to be more consistent with the `NotNullColumnConstraints` arg_type.

File: sqlglot/parser.py (hunk 2 above)
Comment (noklam) [reply sqlglot/parser.py]: Rookie mistake, Thanks!

**Comment by noklam** [comment]
ready to review now,

File: sqlglot/generator.py [hunk 3]
```diff
@@ -3022,6 +3022,10 @@ def altercolumn_sql(self, expression: exp.AlterColumn) -> str:
         if comment:
             return f"ALTER COLUMN {this} COMMENT {comment}"
 
+        allow_null = self.sql(expression, "allow_null")
+        if allow_null:
+            nullability = "SET" if allow_null == "NOT NULL" else "DROP"
+            return f"ALTER COLUMN {this} {nullability} NOT NULL"
```
Comment (georgesittas) [inline sqlglot/generator.py:3028]: What's the point of storing a `exp.NotNullColumnConstraint` in this case, if we're not gonna use the SQL it generates into? I think storing a boolean flag instead will be cleaner, e.g. no string comparison will be required.

File: tests/dialects/test_snowflake.py [hunk 4]
```diff
@@ -1282,6 +1282,13 @@ def test_ddl(self):
             read={"teradata": "CREATE MULTISET TABLE a (b INT)"},
             write={"snowflake": "CREATE TABLE a (b INT)"},
         )
+        self.validate_all(
+            """
+                ALTER TABLE a
+                ALTER COLUMN my_column DROP NOT NULL;
+            """,
+            write={"duckdb": "ALTER TABLE a ALTER COLUMN my_column DROP NOT NULL"},
+        )
```
Comment (georgesittas) [inline tests/dialects/test_snowflake.py:1291]: Make sure to always test the roundtrip, i.e. snowflake -> snowflake. So we need a couple of entries here.

**Comment by noklam** [comment]
I have made changes based on the review

File: sqlglot/generator.py [hunk 5]
```diff
@@ -3022,9 +3022,16 @@ def altercolumn_sql(self, expression: exp.AlterColumn) -> str:
...
             return f"ALTER COLUMN {this} COMMENT {comment}"
 
-        if not expression.args.get("drop"):
+        allow_null = expression.args.get("allow_null")
+        drop = expression.args.get("drop")
+
+        if not drop and not allow_null:
             self.unsupported("Unsupported ALTER COLUMN syntax")
 
+        if allow_null:
```
Comment (georgesittas) [inline sqlglot/generator.py:3031]:
```suggestion
        if allow_null is not None:
```

**Pull request merged**